proposals/p2687.md

```diff
@@ -0,0 +1,298 @@
+`Optional(i32)` but not strictly more complicated than `Optional(bool)`.
+
+This rule, when combined with
+[the acyclic rule](/docs/design/generics/details.md#acyclic-rule) that a query
+can't repeat exactly, [guarantees termination](#proof-of-termination). This rule
+is expected to identify a problematic sequence of `impl` declaration
+instantiations in a way that is easier for the user to understand. Consider the
+example from before,
+
+```
+interface I;
+impl forall [T:! type where Optional(.Self) is I] T as I;
+```
+
+This `impl` declaration matches the query `i32 impls I` as long as
+`Optional(i32) impls I`. That is a strictly more complicated query, though,
+since it contains all the base types of the starting query (`i32` and `I`), plus
+one more (`Optional`). As a result, an error can be given after one step, rather
+than after hitting a large recursion limit. And that error can state explicitly
+what went wrong: we went from a query with no `Optional` to one with one,
+without anything else decreasing.
+
+It is robust in the face of refactoring:
+
+-   It does not depend on the specifics of how an `impl` declaration is
+    parameterized, only on the query.
+-   It does not depend on the length of the chain of queries.
+-   It does not depend on a measure of type-expression complexity, like depth.
+
+## Details
+
+### Non-type arguments
+
+For non-type arguments we have to expand beyond base types to consider other
+kinds of keys. These other keys are in a separate namespace from base types.
+
+-   Values with an integral type use a key with the name of the type and the
+    absolute value as a count. This means integer arguments are considered more
+    complicated if they increase in absolute value. For example, if the values
+    `2` and `-3` are used as arguments to parameters with type `i32`, then the
+    `i32` key will have count `5`.
+-   Every option of a choice type is its own key, counting how many times a
+    value using that option occurs. Any parameters to the option are recorded as
+    separate keys. For example, the `Optional(i32)` value of `.Some(7)` is
+    recorded as keys `.Some` (with a count of `1`) and `i32` (with a count of
+    `7`).
+-   Yet another namespace of keys is used to track counts of variadic arguments,
+    under the base type. This is to defend against having a variadic type `V`
+    that takes any number of `i32` arguments, with an infinite set of distinct
+    instantiations: `V(0)`, `V(0, 0)`, `V(0, 0, 0)`, ...
+    -   A `tuple` key in this namespace is used to track the total number of
+        components of tuple values. The values of those elements will be tracked
+        using their own keys.
+
+Non-type argument values not covered by these cases must not change. That is to
+say, if the 3rd argument to a type constructor `V(...)` is `"foo"`, in one
+query, then it should be the same in every dependent query that includes
+`V(...)`.
```

Comment:
I'm not sure how to apply this rule in the case where the 3rd argument to `V(...)` doesn't exist in the other query at all (because either the `V(...)` is gone or is now in a different place).

I think that, after considering the rest of the rules, we have one of a small number of cases:

1) Some key decreases: the new query is less complex, and is acceptable.
2) No key decreases and some key increases: the new query is more complex, and is rejected.
3) No key decreases or increases, and the query is identical to the one we had before: the new query is a repeat of the earlier one, and is rejected.
4) No key decreases or increases, and the query is not identical to the one we had before: there are only finitely many orderings of the keys, so we accept.

I think what we're looking for here might be, for the "identical to the one we had before" comparison, we completely ignore non-type argument values other than the ones for which we have special-case handling. Put another way, we delete all non-type arguments with no special handling from the query entirely before applying this algorithm.

Reply:
That is a simpler rule, thanks. How is my new text?

Reply:
Oh, I just realized I need to add this as an alternative considered. I'll do that after the meeting.

Reply:
Sorry for the delay. This alternative has now been recorded.